tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir

```diff
@@ -239,3 +239,56 @@ func @main(%arg0 : i32) -> (tensor<i32>, tensor<2x1xi32>) {
   return %0, %1 : tensor<i32>, tensor<2x1xi32>
 }
 
+// -----
+
+// test shape constraint attr attached in op.
+// test symbols dim ops 2/3 are not removed (used in kDiscSymbolicDimAttr)
+
+module {
+  // CHECK-LABEL: @main
+  func @main(%arg0: tensor<?x?xf32, [@S0, @S1]>, %arg1: tensor<?x?xf32, [@S0, @S1]>) -> tensor<?x?xf32, [@S0, @S1]> {
+    %0 = mhlo.add %arg0, %arg1 {kDiscSymbolicDimAttr = [@S2, @S3]} : tensor<?x?xf32, [@S0, @S1]>
+    return %0 : tensor<?x?xf32, [@S0, @S1]>
+  }
+  // Test symbols dim ops 2/3 are not removed (used in kDiscSymbolicDimAttr)
+  // CHECK-DAG: "disc_shape.SymbolicDim"() {knownNegativeOne = false, knownNonNegative = true, knownNonSizeOne = false, knownNonSizeZero = false, sym_name = "S0", value = -1 : i64}
+  // CHECK-DAG: "disc_shape.SymbolicDim"() {knownNegativeOne = false, knownNonNegative = true, knownNonSizeOne = false, knownNonSizeZero = false, sym_name = "S1", value = -1 : i64}
+  // CHECK-DAG: "disc_shape.SymbolicDim"() {knownNegativeOne = false, knownNonNegative = true, knownNonSizeOne = false, knownNonSizeZero = false, sym_name = "S2", value = -1 : i64}
+  // CHECK-DAG: "disc_shape.SymbolicDim"() {knownNegativeOne = false, knownNonNegative = true, knownNonSizeOne = false, knownNonSizeZero = false, sym_name = "S3", value = -1 : i64}
```

Comment:
`S2` and `S3` are not removed in the case?

Reply:
Yes. In this case, they are used by some ops and can not be combined with other symbols as well.